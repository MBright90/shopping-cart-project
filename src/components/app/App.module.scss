.appLayout {
    background-color: #000;
    color: #fff;
    display: flex;
    flex-direction: column;
    line-height: 1;
    min-height: 100vh;
    width: 100vw;
    z-index: -100;
}