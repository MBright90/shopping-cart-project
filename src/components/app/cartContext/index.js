import { object } from 'prop-types'
import React, { createContext, useEffect, useState } from 'react'

const activeVoucherCodes = { mbright90: 0.1, superstarmanson: 0.05 }

export const cartContext = createContext()

const CartContextProvider = ({ children }) => {
  const [cartContents, setCartContents] = useState([])
  const [isShowing, setIsShowing] = useState(false)
  const [activeDiscount, setActiveDiscount] = useState(0)

  const [totalPrice, setTotalPrice] = useState(
    cartContents.length > 0 ? cartContents.reduce((product) => product.price * product.quantity) : 0
  )

  // Sets current total price on any cart contents change
  useEffect(() => {
    console.log(activeDiscount)
    setTotalPrice(
      cartContents.length > 0
        ? cartContents.reduce((total, product) => total + product.price * product.quantity, 0) *
            (1 - activeDiscount)
        : 0
    )
  }, [activeDiscount, cartContents])

  const checkCart = (productId) => {
    if (cartContents.length) return cartContents.find((product) => product.id === productId)
    return false
  }

  const addProductToCart = (product, quantityOrdered) => {
    // Check if an order is already made on this product, if so return false
    // to initiate message to check cart
    if (checkCart(product.id)) return false

    // If no order already for this product, add product and quantity to cart
    const productCopy = { ...product, quantity: quantityOrdered }
    setCartContents([...cartContents, productCopy])
    return true
  }

  const applyDiscountCode = (discountCode) => {
    const discountAppliedFor = activeVoucherCodes[discountCode.toLowercase()]
      ? activeVoucherCodes[discountCode]
      : 0
    if (discountAppliedFor > activeDiscount) {
      setActiveDiscount(discountAppliedFor)
      return true
    }
    return false
  }

  const removeProductFromCart = (productId) => {
    const tempCart = [...cartContents]
    const productIndex = tempCart.findIndex((product) => product.id === productId)
    tempCart.splice(productIndex, 1)
    setCartContents(tempCart)
  }

  const updateCartQuantity = (productId, newQuantity) => {
    const productIndex = cartContents.findIndex((product) => product.id === productId)
    const tempCartContents = [...cartContents]
    tempCartContents[productIndex].quantity = newQuantity
    setCartContents(tempCartContents)
  }

  const contextValue = {
    addProductToCart,
    removeProductFromCart,

    activeDiscount,
    applyDiscountCode,
    cartContents,
    totalPrice,

    isShowing,
    setIsShowing,

    setActiveDiscount,
    updateCartQuantity
  }

  return <cartContext.Provider value={contextValue}>{children}</cartContext.Provider>
}

CartContextProvider.propTypes = {
  children: object.isRequired
}

export default CartContextProvider
