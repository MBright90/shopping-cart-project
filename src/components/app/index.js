import { oneOf } from 'prop-types'

import style from './App.module.scss'
import ProductContext from './productContext'

const App = ({ children }) => {
  return (
    <ProductContext>
      <div className={style.appLayout}>{children}</div>
    </ProductContext>
  )
}

App.propTypes = {
  children: oneOf([Node, NodeList]).isRequired
}

export default App
