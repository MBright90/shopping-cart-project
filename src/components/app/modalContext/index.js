import { object } from 'prop-types'
import React, { createContext, useState } from 'react'

export const modalContext = createContext()

const ModalContextProvider = ({ children }) => {
  const [currentModal, setCurrentModal] = useState(null)

  const contextValue = {
    currentModal,
    setCurrentModal
  }

  return <modalContext.Provider value={contextValue}>{children}</modalContext.Provider>
}

ModalContextProvider.propTypes = {
  children: object.isRequired
}

export default ModalContextProvider
