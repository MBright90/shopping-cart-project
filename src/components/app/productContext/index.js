/* eslint-disable prettier/prettier */
import { object } from 'prop-types'
import React, { createContext, useState } from 'react'

import ampImages from '@assets/images/products/amps'
import guitarImages from '@assets/images/products/guitars'
import pedalImages from '@assets/images/products/pedals'

export const productContext = createContext()

const productDatabase = {
  guitars: [
    {
      name: 'MA-2',
      image: guitarImages.maTwo,
      price: 949,
      description: 'The Manson MA-2 is a versatile and affordable electric guitar that\'s perfect for players of all levels. Its lightweight body and comfortable neck make it easy to play for hours on end, while its high-output pickups deliver a wide range of tones. Whether you\'re playing blues, rock, or metal, the MA-2 can handle it all with ease.'
    },
    {
      name: 'NEBULA-X',
      image: guitarImages.nebula,
      price: 1050,
      description: 'The Manson NEBULA-X is a sleek and futuristic electric guitar that\'s perfect for players who want a modern and innovative instrument. Its unique design and custom pickups deliver a wide range of tones, from crisp and clean to heavy and distorted. Whether you\'re playing on stage or in the studio, the NEBULA-X is sure to turn heads and deliver an unforgettable sound.'
    },
    {
      name: 'DLR-02',
      image: guitarImages.dlr,
      price: 5699,
      description: 'The Manson DLR-02 is a high-end electric guitar that\'s designed for professional players who demand the very best. Its hand-crafted construction, premium materials, and custom pickups deliver a tone that\'s rich, complex, and full of character. Whether you\'re playing jazz, fusion, or experimental music, the DLR-02 is sure to inspire and elevate your playing to new heights.'
    },
    {
      name: 'ORYX VII',
      image: guitarImages.oryx,
      price: 3349,
      description: 'The Manson ORYX VII is a bold and powerful electric guitar that\'s perfect for heavy rock and metal. Its aggressive styling and custom pickups deliver a tone that\'s thick, aggressive, and full of attitude. Whether you\'re playing crushing riffs or blistering solos, the ORYX VII is sure to make a statement and deliver a sound that\'s truly unforgettable.'
    },
    {
      name: 'META MBM-1',
      image: guitarImages.mbmOne,
      price: 499,
      description: 'The Manson META MBM-1 is a signature electric guitar designed in collaboration with Matt Bellamy of the band Muse. Its unique shape, custom pickups, and versatile switching options make it perfect for players who want to replicate Matt\'s iconic sound. Whether you\'re a fan of Muse or just looking for a great guitar that\'s full of character, the META MBM-1 is sure to deliver.'
    },
    {
      name: 'DL-2',
      image: guitarImages.dlTwo,
      price: 1099,
      description: 'The Manson DL-2 is a versatile and high-quality electric guitar that\'s perfect for players who demand the very best. Its lightweight body and custom pickups deliver a wide range of tones, from clean and mellow to heavy and distorted. Whether you\'re playing in a band, recording in the studio, or just jamming at home, the DL-2 is sure to inspire and impress.' 
    }
  ],
  pedals: [
    {
      name: 'Boss DS-1 Distortion Pedal',
      image: pedalImages.bossDs,
      price: 49.99,
      description:
        'The Boss DS-1 Distortion Pedal is a classic distortion pedal used by countless guitarists across the world. With its simple controls and versatile sound, this pedal can deliver anything from a mild overdrive to a heavy metal distortion. It\'s perfect for players who want a reliable, affordable, and easy-to-use distortion pedal.'
    },
    {
      name: 'Electro-Harmonix Big Muff Pi Fuzz Pedal',
      image: pedalImages.bigMuff,
      price: 79.99,
      description:
        'The Electro-Harmonix Big Muff Pi Fuzz Pedal is a legendary fuzz pedal that\'s been used by guitarists for decades. Its unique sound is characterized by a thick, creamy fuzz that\'s perfect for heavy riffs and solos. This pedal features simple controls and is easy to use, making it a great choice for players who want a classic fuzz sound without a lot of tweaking.'
    },
    {
      name: 'TC Electronic Ditto Looper Pedal',
      image: pedalImages.dittoLoop,
      price: 99.99,
      description:
        'The TC Electronic Ditto Looper Pedal is a compact and powerful looper pedal that\'s perfect for guitarists who want to add layers to their playing. With up to five minutes of loop time and unlimited overdubs, this pedal is perfect for solo performers or songwriters who want to build complex arrangements. The Ditto Looper also features intuitive controls and is easy to use, making it a great choice for players of all levels.'
    },
    {
      name: 'MXR M169 Carbon Copy Analog Delay Pedal',
      image: pedalImages.carbonCopy,
      price: 149.99,
      description:
        'The MXR M169 Carbon Copy Analog Delay Pedal is a classic delay pedal that delivers warm, organic delay tones. Its analog circuitry creates a natural, vintage sound that\'s perfect for anything from subtle slapback echoes to long, atmospheric delays. This pedal also features modulation controls that add a unique character to your sound, making it a great choice for players who want a versatile and expressive delay pedal.'
    },
    {
      name: 'Strymon Timeline Delay Pedal',
      image: pedalImages.timelineDelay,
      price: 449.99,
      description:
        'The Strymon Timeline Delay Pedal is a high-end delay pedal that features a wide range of delay sounds and advanced features. With 12 delay machines, 200 presets, and deep editing capabilities, this pedal is perfect for players who want complete control over their delay sound. It also features stereo inputs and outputs, MIDI connectivity, and a powerful looper, making it a versatile and powerful tool for live performance and recording.'
    }
  ],
  amps: [
    {
      name: 'Mesa Boogie Mark V',
      image: ampImages.mesaBoogie,
      price: 2499,
      description: 'The Mesa Boogie Mark V is a versatile and powerful guitar amp that can handle any style of music. With three channels and multiple modes, this amp delivers everything from sparkling cleans to bone-crushing distortion. Whether you\'re playing in a small club or a large arena, the Mark V is sure to provide the tone and power you need to stand out.'
    },
    {
      name: 'Marshall JVM410H',
      image: ampImages.marshallJVM,
      price: 1999,
      description: 'The Marshall JVM410H is a high-end guitar amp that\'s perfect for players who demand the very best. With four channels and multiple modes, this amp delivers a wide range of tones, from classic rock to modern metal. Whether you\'re playing in the studio or on stage, the JVM410H is sure to inspire and impress.'
    },
    {
      name: 'Fender Blues Junior',
      image: ampImages.fenderBlues,
      price: 599,
      description: 'The Fender Blues Junior is a classic and affordable guitar amp that\'s perfect for blues and rock players. With its simple controls and 15 watts of power, this amp delivers a warm and punchy tone that\'s perfect for small gigs and rehearsals. Whether you\'re playing at home or on stage, the Blues Junior is sure to provide the classic Fender sound you love.'
    },
    {
      name: 'Orange Rockerverb 50 MKIII',
      image: ampImages.rockerVerb,
      price: 2099,
      description: 'The Orange Rockerverb 50 MKIII is a high-quality guitar amp that\'s perfect for players who demand the best tone and performance. With its two channels and multiple modes, this amp delivers a wide range of tones, from vintage clean to modern high-gain. Whether you\'re playing in a band or recording in the studio, the Rockerverb 50 MKIII is sure to provide the power and versatility you need.'
    }
  ]
}

const categoryFilters = ['All', 'Guitars', 'Pedals', 'Amps']

const ProductContextProvider = ({ children }) => {
  const [category, setCategory] = useState('all')

  const contextValue = {
    productDatabase,
    category,
    setCategory,
    categoryFilters
  }
  return <productContext.Provider value={contextValue}>{children}</productContext.Provider>
}

ProductContextProvider.propTypes = {
  children: object.isRequired
}

export default ProductContextProvider
