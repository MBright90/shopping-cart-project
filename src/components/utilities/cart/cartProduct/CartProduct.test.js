/* eslint-disable testing-library/no-unnecessary-act */
import { cartContext } from '@components/app/cartContext'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

import CartProduct from './index'

describe('CartProduct', () => {
  let cartValue
  let product

  beforeEach(() => {
    cartValue = {
      removeProductFromCart: () => {},
      updateCartQuantity: () => {}
    }
    product = {
      name: 'testName',
      id: 'testId',
      inStock: true,
      image: 'testImage',
      price: 100,
      description: 'testDescription',
      quantity: 3
    }
  })

  it('renders to match snapshot', () => {
    const { container } = render(<CartProduct product={product} />, {
      wrapper: ({ children }) => (
        <cartContext.Provider value={cartValue}>{children}</cartContext.Provider>
      )
    })
    expect(container).toMatchSnapshot()
  })

  it('calls removeProductFromCart upon delete button click', () => {
    cartValue.removeProductFromCart = jest.fn()

    render(<CartProduct product={product} />, {
      wrapper: ({ children }) => (
        <cartContext.Provider value={cartValue}>{children}</cartContext.Provider>
      )
    })

    const deleteButton = screen.getByTestId('delete-button')
    userEvent.click(deleteButton)

    expect(cartValue.removeProductFromCart).toHaveBeenCalled()
  })

  it('calls updateCartQuantity when quantity buttons are clicked', () => {
    cartValue.updateCartQuantity = jest.fn()

    render(<CartProduct product={product} />, {
      wrapper: ({ children }) => (
        <cartContext.Provider value={cartValue}>{children}</cartContext.Provider>
      )
    })

    const buttons = screen.getAllByRole('button')
    userEvent.click(buttons[0])
    expect(cartValue.updateCartQuantity).toHaveBeenCalled()

    userEvent.click(buttons[1])
    expect(cartValue.updateCartQuantity).toHaveBeenCalledTimes(2)
  })
})
