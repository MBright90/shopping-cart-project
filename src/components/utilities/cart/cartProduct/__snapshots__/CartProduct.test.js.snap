// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProduct renders to match snapshot 1`] = `
<div>
  <div
    class="productCard"
  >
    <div
      class="productDetails"
    >
      <p
        class="productName"
      >
        testName
      </p>
      <p
        class="productPrice"
      >
        £
        100
         
        <span
          class="red"
        >
          each
        </span>
      </p>
    </div>
    <div
      class="productImage"
      style="background-image: url(testImage);"
    />
    <div
      class="productQuantity"
    >
      <div
        class="inputQuantity"
      >
        <button
          class="quantityButton"
          type="button"
        >
          -
        </button>
        <input
          class="quantityInput"
          data-testid="quantity-input"
          id="quantity-input"
          max="9"
          min="0"
          readonly=""
          type="number"
          value="3"
        />
        <button
          class="quantityButton"
          type="button"
        >
          +
        </button>
      </div>
      <p
        class="totalPrice"
      >
        £
        300.00
      </p>
    </div>
    <button
      class="deleteButton"
      data-product-id="testId"
      data-testid="delete-button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-delete-left "
        data-icon="delete-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M576 128c0-35.3-28.7-64-64-64H205.3c-17 0-33.3 6.7-45.3 18.7L9.4 233.4c-6 6-9.4 14.1-9.4 22.6s3.4 16.6 9.4 22.6L160 429.3c12 12 28.3 18.7 45.3 18.7H512c35.3 0 64-28.7 64-64V128zM271 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
